### PR TITLE
feat: (local-video-recordings) enable users to record their conference locally

### PR DESCRIPTION
**Pull request by andioltean**
all users can now record the conference locally
it works in the same way other recordings works
when recording is complete a file is created and stored locally

image

### CODE REVIEW

File: react/features/base/participants/functions.js [hunk 1]
```diff
@@ -475,3 +476,30 @@ export function getRaiseHandsQueue(stateful: Object | Function): Array<Object> {
...
+}
+
+/**
+ * Returns whether the given participant has his hand raised or not.
+ *
+ * @param {Object} participant - The participant.
+ * @returns {boolean} - Whether participant has raise hand or not.
+ */
+export function hasStartedRecording(participant: Object): boolean {
+    const { localRecording } = participant.localRecording;
```
Comment (quitrk) [inline react/features/base/participants/functions.js:501]: this doesn't look right

File: react/features/base/participants/functions.js [hunk 2]
```diff
@@ -475,3 +476,30 @@ export function getRaiseHandsQueue(stateful: Object | Function): Array<Object> {
...
+/**
+ * Get the participants queue with raised hands.
+ *
+ * @param {(Function|Object)} stateful - The (whole) redux state, or redux's
+ * {@code getState} function to be used to retrieve the state
+ * features/base/participants.
+ * @returns {Array<Object>}
+ */
+export function getLocalRecordingQueue(stateful: Object | Function): Array<Object> {
+    const { localRecording } = toState(stateful)['features/base/participants'];
```
Comment (quitrk) [inline react/features/base/participants/functions.js:489]: should this be `localRecordingQueue`?

File: modules/API/API.js [hunk 3]
```diff
@@ -259,6 +262,18 @@ function initCommands() {
             sendAnalytics(createApiEvent('chat.toggled'));
             APP.store.dispatch(toggleChat());
         },
+        'toggle-local-recording': () => {
+            const localParticipant = getLocalParticipant(APP.store.getState());
```
Comment (quitrk) [inline modules/API/API.js:266]: shouldn't this take into account recordings already in progress?

File: react/features/base/participants/middleware.js [hunk 4]
```diff
@@ -587,6 +661,59 @@ function _raiseHandUpdated({ dispatch, getState }, conference, participantId, ne
...
+ * @returns {void}
+ */
+function _localRecordingUpdated({ dispatch, getState }, conference, participantId, newValue) {
+
+    const state = getState();
+
+    dispatch(participantUpdated({
+        conference,
+        id: participantId,
+        localRecording: newValue
```
Comment (quitrk) [inline react/features/base/participants/middleware.js:680]: if this is supposed to be a boolean indicating whether this participant is recording or not, it should be reflected in its naming (initially I thought it is holding some object with recording info)

File: modules/API/API.js [hunk 5]
```diff
@@ -1453,6 +1468,21 @@ class API {
         });
     }
 
+    /**
+     * Notify external application (if API is enabled) that user updated local recording.
+     *
+     * @param {string} id - User id.
+     * @param {string} localRecording - Whether user has raised hand.
```
Comment (quitrk) [inline modules/API/API.js:1475]: why is this a string rather than a boolean? also the naming is confusing

**Review by quitrk — CHANGES_REQUESTED** [review]
it seems that you're using the same `localRecording' naming for different data types (boolean, string, object) and it's creating a lot of confusion

**Comment by damencho** [comment]
I was wondering, is this different than the local recording we have in https://github.com/jitsi/jitsi-meet/tree/master/react/features/local-recording, or these it reuse something from it?

File: react/features/recording/middleware.js [hunk 6]
```diff
@@ -129,6 +146,28 @@ MiddlewareRegistry.register(({ dispatch, getState }) => next => action => {
...
+                    descriptionKey: 'recording.on',
+                    isDismissAllowed: true,
+                    titleKey: 'dialog.recording'
+                };
+
+                dispatch(showNotification(dialogProps, NOTIFICATION_TIMEOUT_TYPE.SHORT));
+            });
+        } else {
+            stopLocalRecording();
+            setTimeout(() => {
```
[the remark this answers is not in the document]
Comment (andioltean) [reply react/features/recording/middleware.js:164]: there's a delay between when the recording is topped and the data is starting to get dumped. I could not find in the docs some event that handled this

**Comment by andioltean** [comment]
Please jenkins test this please

**Comment by andioltean** [comment]
this local video recording

**Comment by andioltean** [comment]
> I was wondering, is this different than the local recording we have in https://github.com/jitsi/jitsi-meet/tree/master/react/features/local-recording, or these it reuse something from it?

this is local video recording

**Comment by damencho** [comment]
Confusing is that they share the same name ...
Even the config is with same name 
https://github.com/jitsi/jitsi-meet/blob/8e9034601d66947840adb21399559f405a80da49/config.js#L865

**Comment by andioltean** [comment]
> Confusing is that they share the same name ... Even the config is with same name
> 
> https://github.com/jitsi/jitsi-meet/blob/8e9034601d66947840adb21399559f405a80da49/config.js#L865

that config is for sound recording only. I didn't add a config for local video recordings. Also I was under the impression that sound recording would not be used aby more.

**Comment by damencho** [comment]
Well that is in the code, and when reading all events for local recording, I think of them as that feature.
Maybe use another name like localVideoRecording

**Comment by monnier** [comment]
Sounds like just the thing I need, but I have some questions:

- You say "when recording is complete a file is created and stored locally".  Is the (partial) recording stored in the file only at the very end?  What happens if the browser/computer crashes after an hour or two of recording?
- Does this require re-encoding the video, and hence loads the local CPU even more?  If so, can the re-encoding be done offline so it doesn't affect the recording itself?

**Comment by saghul** [comment]
Recording is performed using the MediaRecorder API: https://developer.mozilla.org/en-US/docs/Web/API/MediaRecorder -- It basically behaves as an ongoing download. Video needs to be encoded because what we'll capture is the entire tab.

**Comment by ei8fdb** [comment]
Just to check - this local video/audio recording feature will be rolled out to meet.jit.si also? Thanks for working on this. I appreciate it.

**Comment by saghul** [comment]
> Just to check - this local video/audio recording feature will be rolled out to meet.jit.si also? Thanks for working on this. I appreciate it.

Yes it will, once it's ready.

**Comment by gigabyteservice** [comment]
MediaRecorder will only record our own video, does this PR can record videos of all other participants just like jibri? is it a replacement for the jibri?

**Comment by saghul** [comment]
Incorrect. MediaRecorder will record the given mediastream. In this case we feed it a steam which records the current tab, which we obtain with getDisplayMedia.

This is not a replacement for Jibri, but an alternative for simpler use cases.

**Comment by saghul** [comment]
Superseded by https://github.com/jitsi/jitsi-meet/pull/10894

**Pull request closed without merging**